passes/source2il.nim

```diff
@@ -482,6 +482,50 @@ proc exprToIL(c; t: InTree, n: NodeIndex, bu, stmts): SemType =
       c.error("undeclared identifier: " & t.getString(n))
       bu.add Node(kind: IntVal)
       result = prim(tkError)
+  of SourceKind.If:
+    case t.len(n)
+    of 2: # condition and body
+      bu.subTree If:
```

Comment:
So an error is being reported, but it's not fatal, so analysis doesn't stop. So if we want it to continue, which it reasonably should. Meaning that body expression could be treated as `(Drop (TupleCons))` yield the correct `unit` type, while we keep the error report of course.

Is that about right?

Reply:
No wait.. that's not right, I think what you're saying is all this has to live inside `stmt` and not in `expr`. I'm not yet sure what it is, but I don't quite get how to juggling `stmt` vs `expr`.

Reply:
> No wait.. that's not right, I think what you're saying is all this has to live inside `stmt` and not in `expr`

Yep, that's what I'm saying. So instead of `bu.subTree If:`, it should be `stmts.addStmt If:`.

You can think of the `Expr` like a NimSkull `nkStmtListExpr`. The tree in the expression slot (`Expr.expr`, which is built via `bu`) *must* be an `L10` expression, while the trees in the `Expr.stmts` seq (which is accumulated via the `stmts` parameter) must all be `L10` statements.

Reply:
I'm still gathering my thoughts, but what's confusing is that we create and inject a new `bu` inside an `stmts.addStmt`, which gets appended to the `stmts` of the parent all the way up the tree where `exprToIL` started from.

If I do:

```nim
proc exprToIL(c; t: InTree, n: NodeIndex, bu, stmts): SemType =
  ...
  of If:
    stmts.addStmt If:
      # now `bu` isn't about the expression any more, since it appends to `stmts`
      bu.subTree Stmts:
        # I have `bu` for the stmt, but composing in `exprToIL(c, t, n)` calls is awkward, I guess I speculatively pull in both the `Expr.stmts` and the `Expr.expr` fields into the `bu` I have here? like:
        let body = exprToIL(c, t, b)
        bu.add body.stmts
        bu.add body.expr
        # I can't use `inline` as that requires a `stmts`, which is presently the `L10 If`'s statements and not the ones of the freshly minted `Stmts` node's (aka the body position of the `If`)
```

This could be that I'm misunderstanding `stmts`, but from what I can tell it's the part of the tree that's not the last item. I'm tempted to add an `add(bu, e: Expr)` routine that appends the `stmts` and `expr` fields to the `bu`, but I'm wondering if I'm not understanding `inline` and/or how to use the API.

Any thoughts? Apologies that they're not precise questions, my mental model is still really messy.

Reply:
> but what's confusing is that we create and inject a new `bu` inside a `stmts.addStmt`

The idea here was/is that there's always an *active* builder and that it's named `bu`, in order to make it immediately obvious what code does some tree building. Of course, this makes what kind of tree a builder named `bu` creates context-dependent.

For `exprToIL`, there are three overloads:
1. the external one for creating a procedure out of an expression (`proc(c; t: InTree)`)
2. the overload that implements the core of the expression analysis and translation (`proc(c; t: InTree; n: NodeIndex, bu, stmts): SemType`)
3. the wrapper overload (`proc(c; t: InTree, n: NodeIndex): Expr`)

In most cases, if not all, the overload you want to call is the third one. Using the second one only makes sense to use as an optimization, for when you want to directly append to a `stmts` and `bu`.

When you do `addStmt`, a builder is provided to the code block for building the statement tree, which is then appended to the leading statements of the current expressions (if `stmts` is the parameter of the `exrpToIL` procedure).  

> but from what I can tell it's the part of the tree that's not the last item

Broadly speaking, yes, but do note that `stmts` is a `seq` (not just a single tree) where each item is a tree representing an `L10` statement.

> If I do:

You're pseudo-code is almost correct, except for the `bu.add body.expr`. All direct children of a `Stmts` tree must be *statements*, so adding an `Expr.expr` (which is an expression, or at least it will be, with #49) results in wrong syntax.

This is also why there's no `inline(bu; e: Expr)` -- there's no context in which adding the statements and expression to a builder immediately after each other is valid.

What would work is adding a `genAsgn(..., e: Expr, bu)` overload.

> but I'm wondering if I'm not understanding `inline` and/or how to use the API.

`inline` is meant for simple unary operator expressions (see `notToIL` for an example), where the operator is applied directly to the argument expression. Maybe the name is confusing.

Reply:
> Broadly speaking, yes, but do note that `stmts` is a `seq` (not just a single tree) where each item is a tree representing an `L10` statement.

This seems to have clicked for me, when I was thinking of composition I wasn't keeping in mind that I'm composing lowered things, so I'm not smashing source expressions, which can be statements (as in `void` or `unit`), but `L10`.

The other habit lying around from the Skull compiler is that whenever I see Skull code I immediately assume low trust and was worried some `expr`, or `stmts`, where going to be forgotten, rather than trusting the invariant will be maintained in `exprToIL` to no return illformed things.

Thanks